### ISSUE TITLE
Spread operator should declare members as undefined for null or undefined objects   

### ISSUE THREAD

**Issue by stephanedr** [Microsoft/TypeScript#13066]
**TypeScript Version:**  2.1.4 & 2.2.0-dev.20161220 

**Code**
```ts
function test(t?: { a: number } | null): void {
    let b = { ...t };  // b: { a: number }
    let c: number = b.a;  // no error
}
```

**Expected behavior:**
`b.a` to be `number | undefined`, as if `t` is `null` or `undefined`, `b` is an empty object.
**Actual behavior:**
`b.a`is `number`.

**Comment by andy-ms** [comment]
This can also cause a value of a completely wrong type to be accepted:
```ts
interface I {
    "foo-bar": "foo" | "bar";
}

function f(x: I | undefined): I {
    return { "foo-bar": 123, ...x }
}

console.log(f(undefined));
```

**Comment by sandersn** [comment]
`getSpreadType` filtered `null` and `undefined` before lifting unions. It should wait until after unions are lifted so that, for example the type of the spread in @andy-ms's example is `{ "foo-bar": number | string }`. I should have a fix out soon.

**Comment by sandersn** [comment]
Fix is up at #16306
